### PR TITLE
Backend implementation

### PR DESCRIPTION
**Pull request by alindima**
This PR implements the Rust IR and the logic for compiling it into BPF code.

In order for the design to be fully implemented, we need to add the JSON frontend, which will come in a future PR.

### CODE REVIEW

File: src/backend/bpf.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+// Builds a `jump` BPF instruction.
+//
+// # Arguments
+//
+// * `code` - The operation code.
+// * `jt` - The jump offset in case the operation returns `true`.
+// * `jf` - The jump offset in case the operation returns `false`.
+// * `k` - The operand.
+#[allow(non_snake_case)]
```
Comment (andreeaflorescu) [inline src/backend/bpf.rs:47]: What is the reason for naming these functions with upper case? They are not const.

File: src/backend/mod.rs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        match self {
+            TargetArch::x86_64 => AUDIT_ARCH_X86_64,
+            TargetArch::aarch64 => AUDIT_ARCH_AARCH64,
+        }
+    }
+}
+
+impl TryFrom<&str> for TargetArch {
+    type Error = Error;
+    fn try_from(input: &str) -> Result<Self> {
```
Comment (andreeaflorescu) [inline src/backend/mod.rs:94]: Why is this needed?

File: src/backend/mod.rs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+pub enum TargetArch {
+    /// x86_64 arch
+    x86_64,
+    /// aarch64 arch
+    aarch64,
+}
+
+impl TargetArch {
+    /// Get the arch audit value. Used for the runtime arch check embedded in the BPF filter.
+    fn get_audit_value(self) -> u32 {
```
Comment (andreeaflorescu) [inline src/backend/mod.rs:84]: This function is not needed because enum values are numeric types. We can instead define `TargetArch` as follows:

```rust
#[allow(non_camel_case_types)]
#[derive(Debug, PartialEq, Clone, Copy)]
pub enum TargetArch {
    /// x86_64 arch
    x86_64 = AUDIT_ARCH_X86_64 as isize,
    /// aarch64 arch
    aarch64 = AUDIT_ARCH_AARCH64 as isize,
}
```

And then use its value directly (or convert it to other numeric types if needed).

File: src/lib.rs [hunk 4]
```diff
@@ -1,6 +1,210 @@
...
+
+/// Library errors.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Error originating in the backend compiler.
+    Backend(BackendError),
+    /// Attempting to install an empty filter.
+    EmptyFilter,
+    /// System error related to calling `prctl`.
+    Prctl(i32),
```
Comment (andreeaflorescu) [inline src/lib.rs:132]: This makes it harder to read the errors. Why aren't we using `std::io`? If we need the `PartialEq` which is a good reason in itself, we should improve the error message and say that the arbitrary number is an errno.

File: src/backend/filter.rs [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        let chain: Vec<_> = chain
+            .into_iter()
+            .map(|rule| {
+                let mut bpf: BpfProgram = rule.into();
+                // Last statement is the on-match action of the filter.
+                bpf.push(BPF_STMT(BPF_RET | BPF_K, u32::from(filter_action.clone())));
+                bpf
+            })
+            .collect();
+        let chain_len: usize = chain.iter().map(std::vec::Vec::len).sum();
```
Comment (andreeaflorescu) [inline src/backend/filter.rs:132]: nit: `Vec::len` instead of std::vec::...

File: src/backend/filter.rs [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+
+        // The chain starts with a comparison checking the loaded syscall number against the
+        // syscall number of the chain.
+        let mut built_syscall = Vec::with_capacity(chain_len + 2);
+        built_syscall.push(BPF_JUMP(
+            BPF_JMP | BPF_JEQ | BPF_K,
+            // Safe because linux system call numbers are nowhere near the u32::MAX value.
```
Comment (andreeaflorescu) [inline src/backend/filter.rs:139]: Why do we define them as `i64` then?

File: src/backend/bpf.rs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        jt: 0,
+        jf: 0,
+        k,
+    }
+}
+
+// Builds a sequence of BPF instructions that validate the underlying architecture.
+#[allow(non_snake_case)]
+#[inline(always)]
+pub(crate) fn VALIDATE_ARCHITECTURE(target_arch: TargetArch) -> Vec<sock_filter> {
```
Comment (andreeaflorescu) [inline src/backend/bpf.rs:73]: nit: the name is misleading. It sounds like something gets validated, but the function actually builds a architecture validator. Can you rename it?

File: src/backend/bpf.rs [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+                code: 0x15,
+                jt: 2,
+                jf: 5,
+                k: 10,
+            }
+        );
+    }
+
+    #[test]
+    fn test_validate_architecture() {
```
Comment (andreeaflorescu) [inline src/backend/bpf.rs:184]: This test function just duplicates the code. Can we remove it?

File: tests/integration_tests.rs [hunk 9]
```diff
@@ -0,0 +1,776 @@
...
+        || unsafe {
+            libc::ioctl(0, 0);
+        },
+        Some(true),
+    );
+
+    // check use cases for QWORD
+    let rules = vec![(
+        libc::SYS_ioctl,
+        vec![SeccompRule::new(vec![Cond::new(2, Qword, Eq, std::u64::MAX).unwrap()]).unwrap()],
```
Comment (andreeaflorescu) [inline tests/integration_tests.rs:180]: std::u64::MAX is deprecated. Can you replace it with `u64::MAX`?

**Review by andreeaflorescu — COMMENTED** [review]
Did an initial pass. I still need to take a closer look at the tests.

File: coverage_config_x86_64.json [hunk 10]
```diff
@@ -1,5 +1,5 @@
 {
-    "coverage_score": 0,
+    "coverage_score": 96.8,
```
Comment (andreeaflorescu) [inline coverage_config_x86_64.json:2]: We should exclude the `tests/integration_tests.rs` path, otherwise it counts towards the coverage.

File: tests/integration_tests.rs (hunk 9 above)
Comment (alindima) [reply tests/integration_tests.rs:180]: didn't know that, sure

File: src/lib.rs [hunk 11]
```diff
@@ -1,6 +1,210 @@
...
+/// * `bpf_filter` - A reference to the [`BpfProgram`] to be installed.
+///
+/// [`BpfProgram`]: struct.BpfProgram.html
+pub fn apply_filter(bpf_filter: BpfProgramRef) -> Result<()> {
+    // If the program is empty, don't install the filter.
+    if bpf_filter.is_empty() {
+        return Err(Error::EmptyFilter);
+    }
+
+    unsafe {
```
[the remark this answers is not in the document]
Comment (alindima) [reply src/lib.rs:166]: Good suggestion! I added comments to the `unsafe` blocks and tried to keep them as small as possible.

In regards to using `std::io::Error::las_os_error()`, I've seen it not reporting errors when the syscall was not io-related.
However, I switched to the `std::io::Error::from_raw_os_error(rc)` variant, which works as expected.

File: src/lib.rs (hunk 4 above)
Comment (alindima) [reply src/lib.rs:132]: good idea. Replaced this with `std::io::Error`. The `PartialEq` is not a hard requirement. It was just useful when testing but it can be worked around

File: coverage_config_x86_64.json (hunk 10 above)
Comment (alindima) [reply coverage_config_x86_64.json:2]: 👍🏻  updated it

File: src/backend/filter.rs [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+    ///     std::env::consts::ARCH.try_into().unwrap(),
+    /// );
+    /// ```
+    ///
+    /// [`SeccompRule`]: struct.SeccompRule.html
+    /// [`SeccompAction`]: struct.SeccompAction.html
+    pub fn new(
+        rules: BTreeMap<i64, Vec<SeccompRule>>,
+        default_action: SeccompAction,
+        filter_action: SeccompAction,
```
[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply src/backend/filter.rs:85]: The value of renaming the actions comes from renaming both of them so that it is intuitive in respect to where are they applies (on match or on mis-match). I prefer us renaming both of them.

What other seccomp libraries have you looked at out of curiosity?

File: src/backend/bpf.rs (hunk 1 above)
Comment (andreeaflorescu) [reply src/backend/bpf.rs:50]: I don't think that keeping the upper case name helps in this case. The reason it is written in uppercase in the kernel is that it is a macro, and the C convention says that macros are all caps. Since this is Rust code, we should follow the Rust convention.

When reading the code you are misguided because what seems to be a constant, is actually a function.

File: src/backend/mod.rs (hunk 2 above)
Comment (andreeaflorescu) [reply src/backend/mod.rs:94]: I didn't see it used anywhere. Is this for users of the library?

File: src/lib.rs [hunk 13]
```diff
@@ -1,6 +1,193 @@
...
+    // If the program is empty, don't install the filter.
+    if bpf_filter.is_empty() {
+        return Err(Error::EmptyFilter);
+    }
+
+    // Safe because syscall arguments are valid.
+    let rc = unsafe { libc::prctl(libc::PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0) };
+    if rc != 0 {
+        return Err(Error::Prctl(io::Error::from_raw_os_error(rc)));
+    }
```
Comment (andreeaflorescu) [inline src/lib.rs:171]: According to the documentation this is not the correct way of retrieving the error:
```
On error,
       -1 is returned, and errno is set to indicate the error.
```
https://man7.org/linux/man-pages/man2/prctl.2.html

Comment (alindima) [reply src/lib.rs:171]: ah yes, silly mistake. I'll fix this

**Comment by likebreath** [comment]
@alindima Thank you for the great work. I finished a [draft](https://github.com/likebreath/cloud-hypervisor/commit/531b1523a5d3c02bd5d53e72e010698245a5046c) of integrating Cloud Hypervisor with your PR, and can confirm it works as expected with both SeccompAction Trap and Log. 

Overall, I think the new interface (APIs and Types) are pretty good and cleaner. Admittedly it requires lots of changes due to the dramatic interface changes.

Only one comment/open-question: I wonder whether we should include some [similar macros](https://github.com/likebreath/cloud-hypervisor/blob/531b1523a5d3c02bd5d53e72e010698245a5046c/vmm/src/seccomp_filters.rs#L20-L37) to make it easier (or more readable) to define complex `SeccompCondition`, for example the `kvm ioctls` [rules](https://github.com/likebreath/cloud-hypervisor/blob/531b1523a5d3c02bd5d53e72e010698245a5046c/vmm/src/seccomp_filters.rs#L269) in Cloud Hypervisor.

**Review by likebreath — APPROVED** [review]

File: src/backend/filter.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (alindima) [reply src/backend/filter.rs:85]: libseccomp & libseccomp golang use `default_action` and `action`.
minijail does not even allow the user to specify the actions. It uses `allow` for on-match and kill or trap for mismatch (depending on what is available on the host)

File: src/backend/bpf.rs (hunk 1 above)
Comment (alindima) [reply src/backend/bpf.rs:50]: I think there is no right or wrong in this case. Really I'm fine either way

File: src/backend/mod.rs (hunk 2 above)
Comment (alindima) [reply src/backend/mod.rs:94]: Yes, it's mostly for the users of the library. It's also used in tests

File: src/backend/filter.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (alindima) [reply src/backend/filter.rs]: I'm fine either way, so I renamed the actions to `match_action` and `mismatch_action`, per your suggestion

File: src/backend/bpf.rs (hunk 1 above)
Comment (alindima) [reply src/backend/bpf.rs]: I've lowercased the names

File: src/lib.rs (hunk 13 above)
Comment (alindima) [reply src/lib.rs]: fixed

**Comment by alindima** [comment]
> @alindima Thank you for the great work. I finished a [draft](https://github.com/likebreath/cloud-hypervisor/commit/531b1523a5d3c02bd5d53e72e010698245a5046c) of integrating Cloud Hypervisor with your PR, and can confirm it works as expected with both SeccompAction Trap and Log.

Thank you @likebreath for your work on integrating seccompiler in Cloud hypervisor, it's great to hear it works as expected!
I've taken a quick look over your draft PR and it looks good.

> Overall, I think the new interface (APIs and Types) are pretty good and cleaner. Admittedly it requires lots of changes due to the dramatic interface changes.

Great to hear the interface has been indeed improved. Unfortunately it does require quite a lot of changes, but most of them are repetitive (and automatable with regex find&replace) and the one-time character of the changes I think make the effort worth it 😄 

> Only one comment/open-question: I wonder whether we should include some [similar macros](https://github.com/likebreath/cloud-hypervisor/blob/531b1523a5d3c02bd5d53e72e010698245a5046c/vmm/src/seccomp_filters.rs#L20-L37) to make it easier (or more readable) to define complex `SeccompCondition`, for example the `kvm ioctls` [rules](https://github.com/likebreath/cloud-hypervisor/blob/531b1523a5d3c02bd5d53e72e010698245a5046c/vmm/src/seccomp_filters.rs#L269) in Cloud Hypervisor.

That is a good question. I've given the issue some thought. The two macros have a pretty small implementation and are quite an opinionated way of instantiating the objects. For example, other users of the library may not want to `unwrap()` the result of `SeccompRule::new(vec![$($x),*])`, like you are doing in your integration.

Given the above, I think it would be fine to keep those small macros in CH for now. What do you think?

Of course, we can revisit this choice later on if we come up with better abstractions that make the library easier to use for everybody.

**Comment by likebreath** [comment]
> That is a good question. I've given the issue some thought. The two macros have a pretty small implementation and are quite an opinionated way of instantiating the objects. For example, other users of the library may not want to `unwrap()` the result of `SeccompRule::new(vec![$($x),*])`, like you are doing in your integration.

Right. That's something I wanted to improve hopefully at the time of landing the integration to CH (after landing your PR of course).

> Given the above, I think it would be fine to keep those small macros in CH for now. What do you think?

This sounds good to me.

> Of course, we can revisit this choice later on if we come up with better abstractions that make the library easier to use for everybody.

 Agree. Improving abstractions over creating complex conditions definitely worth a revisit.

File: tests/integration_tests.rs [hunk 14]
```diff
@@ -0,0 +1,785 @@
...
+    libc::SYS_munmap,
+    libc::SYS_exit,
+    libc::SYS_rt_sigreturn,
+    libc::SYS_futex,
+];
+
+fn validate_seccomp_filter(
+    rules: Vec<(i64, Vec<SeccompRule>)>,
+    validation_fn: fn(),
+    should_fail: Option<bool>,
```
Comment (andreeaflorescu) [inline tests/integration_tests.rs:40]: nit: It is a bit hard to understand how to use this parameter, specifically what is the meaning of `None`. We could define an enum in which we have 3 options (with some better names maybe):

```rust
enum Return {
    Errno(u32),
    NoErrno,
    NA,
}
```

**Review by andreeaflorescu — APPROVED** [review]
There are still a few `std::u32::max` calls that should be replaced with `u32::max`.

**Review by sboeuf — APPROVED** [review]
